staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -135,6 +177,11 @@ func (i *indexedWatchers) addWatcher(w *cacheWatcher, number int, value string,
 	} else {
 		i.allWatchers.addWatcher(w, number)
 	}
+
+	// Add it to time bucket only when client support watch bookmarks.
+	if w.allowWatchBookmarks {
```

Comment:
We probably need to check also if feature gate is switched on - otherwise we may be adding watcher to the map, but never remove...

Reply:
Or rather (based on the comment from above):
- I would move this explictly to the  Watch function to the last func() in it:
```
c.watchers.addWatcher(...)
if c.watchBookmarksEnabled && watcher.allowWatchBookmarks {
  c.bookmarkWatchers.addWatcher(w)
}
```

[watchBookmarksEnabled initialized based on feature gate, to avoid accessing feature-gate map every time]